man/man8/pwconv.8
'\" t
.\"     Title: pwconv
.\"    Author: Marek Michałkiewicz
.\" Generator: DocBook XSL Stylesheets v1.78.0 <http://docbook.sf.net/>
.\"      Date: 10/15/2014
.\"    Manual: System Management Commands
.\"    Source: shadow-utils 4.2
.\"  Language: English
.\"
.TH "PWCONV" "8" "10/15/2014" "shadow\-utils 4\&.2" "System Management Commands"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
pwconv, pwunconv, grpconv, grpunconv \- convert to and from shadow passwords and groups
.SH "SYNOPSIS"
.HP \w'\fBpwconv\fR\ 'u
\fBpwconv\fR [\fIoptions\fR]
.HP \w'\fBpwunconv\fR\ 'u
\fBpwunconv\fR [\fIoptions\fR]
.HP \w'\fBgrpconv\fR\ 'u
\fBgrpconv\fR [\fIoptions\fR]
.HP \w'\fBgrpunconv\fR\ 'u
\fBgrpunconv\fR [\fIoptions\fR]
.SH "DESCRIPTION"
.PP
The
\fBpwconv\fR
command creates
\fIshadow\fR
from
\fIpasswd\fR
and an optionally existing
\fIshadow\fR\&.
.PP
The
\fBpwunconv\fR
command creates
\fIpasswd\fR
from
\fIpasswd\fR
and
\fIshadow\fR
and then removes
\fIshadow\fR\&.
.PP
The
\fBgrpconv\fR
command creates
\fIgshadow\fR
from
\fIgroup\fR
and an optionally existing
\fIgshadow\fR\&.
.PP
The
\fBgrpunconv\fR
command creates
\fIgroup\fR
from
\fIgroup\fR
and
\fIgshadow\fR
and then removes
\fIgshadow\fR\&.
.PP
These four programs all operate on the normal and shadow password and group files:
/etc/passwd,
/etc/group,
/etc/shadow, and
/etc/gshadow\&.
.PP
Each program acquires the necessary locks before conversion\&.
\fBpwconv\fR
and
\fBgrpconv\fR
are similar\&. First, entries in the shadowed file which don\*(Aqt exist in the main file are removed\&. Then, shadowed entries which don\*(Aqt have `x\*(Aq as the password in the main file are updated\&. Any missing shadowed entries are added\&. Finally, passwords in the main file are replaced with `x\*(Aq\&. These programs can be used for initial conversion as well to update the shadowed file if the main file is edited by hand\&.
.PP

\fBpwconv\fR
will use the values of
\fIPASS_MIN_DAYS\fR,
\fIPASS_MAX_DAYS\fR, and
\fIPASS_WARN_AGE\fR
from
/etc/login\&.defs
when adding new entries to
/etc/shadow\&.
.PP
Likewise
\fBpwunconv\fR
and
\fBgrpunconv\fR
are similar\&. Passwords in the main file are updated from the shadowed file\&. Entries which exist in the main file but not in the shadowed file are left alone\&. Finally, the shadowed file is removed\&. Some password aging information is lost by
\fBpwunconv\fR\&. It will convert what it can\&.
.SH "OPTIONS"
.PP
The options which apply to the
\fBpwconv\fR,
\fBpwunconv\fR,
\fBgrpconv\fR, and
\fBgrpunconv\fR
commands are:
.PP
\fB\-h\fR, \fB\-\-help\fR
.RS 4
Display help message and exit\&.
.RE
.PP
\fB\-R\fR, \fB\-\-root\fR\ \&\fICHROOT_DIR\fR
.RS 4
Apply changes in the
\fICHROOT_DIR\fR
directory and use the configuration files from the
\fICHROOT_DIR\fR
directory\&.
.RE
.SH "BUGS"
.PP
Errors in the password or group files (such as invalid or duplicate entries) may cause these programs to loop forever or fail in other strange ways\&. Please run
\fBpwck\fR
and
\fBgrpck\fR
to correct any such errors before converting to or from shadow passwords or groups\&.
.SH "CONFIGURATION"
.PP
The following configuration variable in
/etc/login\&.defs
changes the behavior of
\fBgrpconv\fR
and
\fBgrpunconv\fR:
.PP
\fBMAX_MEMBERS_PER_GROUP\fR (number)
.RS 4
Maximum members per group entry\&. When the maximum is reached, a new group entry (line) is started in
/etc/group
(with the same name, same password, and same GID)\&.
.sp
The default value is 0, meaning that there are no limits in the number of members in a group\&.
.sp
This feature (split group) permits to limit the length of lines in the group file\&. This is useful to make sure that lines for NIS groups are not larger than 1024 characters\&.
.sp
If you need to enforce such limit, you can use 25\&.
.sp
Note: split groups may not be supported by all tools (even in the Shadow toolsuite)\&. You should not use this variable unless you really need it\&.
.RE
.PP
The following configuration variables in
/etc/login\&.defs
change the behavior of
\fBpwconv\fR:
.PP
\fBPASS_MAX_DAYS\fR (number)
.RS 4
The maximum number of days a password may be used\&. If the password is older than this, a password change will be forced\&. If not specified, \-1 will be assumed (which disables the restriction)\&.
.RE
.PP
\fBPASS_MIN_DAYS\fR (number)
.RS 4
The minimum number of days allowed between password changes\&. Any password changes attempted sooner than this will be rejected\&. If not specified, \-1 will be assumed (which disables the restriction)\&.
.RE
.PP
\fBPASS_WARN_AGE\fR (number)
.RS 4
The number of days warning given before a password expires\&. A zero means warning is given only upon the day of expiration, a negative value means no warning is given\&. If not specified, no warning will be provided\&.
.RE
.SH "FILES"
.PP
/etc/login\&.defs
.RS 4
Shadow password suite configuration\&.
.RE
.SH "SEE ALSO"
.PP

\fBgrpck\fR(8),
\fBlogin.defs\fR(5),
\fBpwck\fR(8)\&.


NEWS
$Id$

shadow-4.2 -> shadow-4.2.1					UNRELEASED

*** general
  * Properly release with a generated configure script, built manpages
  * Include the configure_userns patch from Debian

*** translation
  * Vietnamese translation update
  * French man pages translation update
  * French translation update
  * Typo fixes in German translation

shadow-4.1.5.1 -> shadow-4.2					2013-04-30

*** general
  * Handle libc whose crypt() returns NULL when passed a salt that
    violates specs or system requirements (e.g. FIPS140). This is needed
    with glibc/eglibc 2.17 for tools checking passwords (passwd (non PAM
    enabled) or newgrp), and for tools generating encrypted passwords
    (chgpasswd, chpasswd, or gpasswd when non PAM enabled or when a fixed
    crypt method is requested on the command line, and newusers, or passwd
    in their non PAM enabled versions)
  * Fix segfault when reading groups split on multiple lines. This impacts
    most user/group management tools when MAX_MEMBERS_PER_GROUP is set.

- su
  * When su receives a signal (SIGTERM, or SIGINT/SIGQUIT in non
    interactive mode), kill the child process group, rather than just the 
    immediate child.
  * Fix segmentation faults for users without a proper home or shell in
    their passwd entries.

- login
  * Fix segmentation faults for users without a proper home or shell in
    their passwd entries.

*** documentation
  * Fixed useradd man page (--home-dir option, instead of --home).

*** translation
  * Updated Russian translation.
  * Updated German man pages translation.
  * Fixed gshadow Japanese man page translation.

shadow-4.1.5 -> shadow-4.1.5.1					2012-05-25

- login
  * Log into utmp(x) when PAM is enabled, but do not log into wtmp.
    This complete pam_lastlog which logs into wtmp and in into utmp(x).
- su
  * non PAM enabled versions: do not fail if su is called without a
    controlling terminal.
- userdel
  * Fix segfault when userdel removes the user's group.

*** documentation
  * .so links now point to paths relative to the top-level manual hierarchy

*** translation
  * Updated French man pages translation.
  * Updated German man pages translation.
  * Updated Polish man pages translation. (logoutd.8)

shadow-4.1.4.3 -> shadow-4.1.5					2012-02-12

*** security
  * su -c could be abused by the executed command to invoke commands with
    the caller privileges. See below. (CVE-2005-4890)

*** general
  * report usage error to stderr, but report usage help to stdout (and return
    zero) when explicitly requested (e.g. with --help).
  * initial support for tcb (http://openwall.com/tcb/) for useradd,
    userdel, usermod, chage, pwck, vipw.
  * Added support for ACLs and Extended Attributes in useradd and usermod.
    Support shall be enabled with the new --with-acl or --with-attr
    configure options.
  * Added diagnosis for lock failures.
  * use libsemanage instead of the semanage tool.

- chage
  * Add --root option.
- chfn
  * Add --root option.
- chgpasswd
  * When the gshadow file exists but there are no gshadow entries, an entry
    is created if the password is changed and group requires a
    shadow entry.
  * Add --root option.
- chpasswd
  * PAM enabled versions: restore the -e option to allow restoring
    passwords without knowing those passwords. Restore together the -m
    and -c options. (These options were removed in shadow-4.1.4 on PAM
    enabled versions)
  * When the shadow file exists but there are no shadow entries, an entry
    is created if the password is changed and passwd requires a
    shadow entry.
  * Add --root option.
- chsh
  * Add --root option.
- faillog
  * The -l, -m, -r, -t options only act on the existing users, unless -a is
    specified.
  * Add --root option.
- gpasswd
  * Add --root option.
- groupadd
  * Add --root option.
- groupdel
  * Add --root option.
- groupmems
  * Fix parsing of gshadow entries.
  * Add --root option.
- groupmod
  * Fixed groupmod when configured with --enable-account-tools-setuid.
  * When the gshadow file exists but there are no gshadow entries, an entry
    is created if the password is changed and group requires a
    shadow entry.
  * Add --root option.
- grpck
  * Add --root option.
  * NIS entries were dropped by -s (sort).
- grpconv
  * Add --root option.
- grpunconv
  * Add --root option.
- lastlog
  * Add --root option.
- login
  * Fixed limits support (non PAM enabled versions only)
  * Added support for infinite limits and group based limits (non PAM
    enabled versions only)
  * Fixed infinite loop when CONSOLE is configured with a colon-separated
    list of TTYs.
  * Fixed warning and support for CONSOLE_GROUPS for users member of more
    than 16 groups.
  * Do not log into utmp(x) or wtmp when PAM is enabled. This is done by
    pam_lastlog.
- newgrp, sg
  * Fix parsing of gshadow entries.
- newusers
  * Add --root option.
- passwd
  * Add --root option.
- pwpck
  * NIS entries were dropped by -s (sort).
  * Add --root option.
- pwconv
  * Add --root option.
- pwunconv
  * Add --root option.
- useradd
  * If the skeleton directory contained hardlinked files, copies of the
    hardlink were removed from the skeleton directory.
  * Add --root option.
- userdel
  * Check the existence of the user's mail spool before trying to remove
    it. If it does not exist, a warning is issued, but no failure.
  * Do not remove a group with the same name as the user (usergroup) if
    this group isn't the user's primary group.
  * Add --root option.
  * Add --selinux-user option.
- usermod
  * Accept options in any order (username not necessarily at the end)
  * When the shadow file exists but there are no shadow entries, an entry
    is created if the password is changed and passwd requires a
    shadow entry, or if aging features are used (-e or -f).
  * Add --root option.
- su
  * Document the su exit values.
  * When su receives a signal, wait for the child to terminate (after

src/pwconv.c
/*
 * Copyright (c) 1996 - 2000, Marek Michałkiewicz
 * Copyright (c) 2002 - 2006, Tomasz Kłoczko
 * Copyright (c) 2009 - 2012, Nicolas François
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the copyright holders or contributors may not be used to
 *    endorse or promote products derived from this software without
 *    specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT
 * HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * pwconv - create or update /etc/shadow with information from
 * /etc/passwd.
 *
 * It is more like SysV pwconv, slightly different from the original Shadow
 * pwconv. Depends on "x" as password in /etc/passwd which means that the
 * password has already been moved to /etc/shadow. There is no need to move
 * /etc/npasswd to /etc/passwd, password files are updated using library
 * routines with proper locking.
 *
 * Can be used to update /etc/shadow after adding/deleting users by editing
 * /etc/passwd. There is no man page yet, but this program should be close
 * to pwconv(1M) on Solaris 2.x.
 *
 * Warning: make sure that all users have "x" as the password in /etc/passwd
 * before running this program for the first time on a system which already
 * has shadow passwords. Anything else (like "*" from old versions of the
 * shadow suite) will replace the user's encrypted password in /etc/shadow.
 *
 * Doesn't currently support pw_age information in /etc/passwd, and doesn't
 * support DBM files. Add it if you need it...
 *
 */

#include <config.h>

#ident "$Id: pwconv.c 3743 2012-05-25 11:51:53Z nekral-guest $"

#include <errno.h>
#include <fcntl.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <getopt.h>
#include "defines.h"
#include "getdef.h"
#include "prototypes.h"
#include "pwio.h"
#include "shadowio.h"
#include "nscd.h"

/*
 * exit status values
 */
/*@-exitarg@*/
#define E_SUCCESS	0	/* success */
#define E_NOPERM	1	/* permission denied */
#define E_USAGE		2	/* invalid command syntax */
#define E_FAILURE	3	/* unexpected failure, nothing done */
#define E_MISSING	4	/* unexpected failure, passwd file missing */
#define E_PWDBUSY	5	/* passwd file(s) busy */
#define E_BADENTRY	6	/* bad shadow entry */
/*
 * Global variables
 */
const char *Prog;

static bool spw_locked = false;
static bool pw_locked = false;

/* local function prototypes */
static void fail_exit (int status);
static void usage (int status);
static void process_flags (int argc, char **argv);

static void fail_exit (int status)
{
	if (pw_locked) {
		if (pw_unlock () == 0) {
			fprintf (stderr, _("%s: failed to unlock %s\n"), Prog, pw_dbname ());
			SYSLOG ((LOG_ERR, "failed to unlock %s", pw_dbname ()));
			/* continue */
		}
	}

	if (spw_locked) {
		if (spw_unlock () == 0) {
			fprintf (stderr, _("%s: failed to unlock %s\n"), Prog, spw_dbname ());
			SYSLOG ((LOG_ERR, "failed to unlock %s", spw_dbname ()));
			/* continue */
		}
	}

	exit (status);
}

static void usage (int status)
{
	FILE *usageout = (E_SUCCESS != status) ? stderr : stdout;
	(void) fprintf (usageout,
	                _("Usage: %s [options]\n"
	                  "\n"
	                  "Options:\n"),
	                Prog);
	(void) fputs (_("  -h, --help                    display this help message and exit\n"), usageout);
	(void) fputs (_("  -R, --root CHROOT_DIR         directory to chroot into\n"), usageout);
	(void) fputs ("\n", usageout);
	exit (status);
}

/*
 * process_flags - parse the command line options
 *
 *	It will not return if an error is encountered.
 */
static void process_flags (int argc, char **argv)
{
	/*
	 * Parse the command line options.
	 */
	int c;
	static struct option long_options[] = {
		{"help", no_argument,       NULL, 'h'},
		{"root", required_argument, NULL, 'R'},
		{NULL, 0, NULL, '\0'}
	};

	while ((c = getopt_long (argc, argv, "hR:",
	                         long_options, NULL)) != -1) {
		switch (c) {
		case 'h':
			usage (E_SUCCESS);
			/*@notreached@*/break;
		case 'R': /* no-op, handled in process_root_flag () */
			break;
		default:
			usage (E_USAGE);
		}
	}

	if (optind != argc) {
		usage (E_USAGE);
	}
}

int main (int argc, char **argv)
{
	const struct passwd *pw;
	struct passwd pwent;
	const struct spwd *sp;
	struct spwd spent;

	Prog = Basename (argv[0]);

	(void) setlocale (LC_ALL, "");
	(void) bindtextdomain (PACKAGE, LOCALEDIR);
	(void) textdomain (PACKAGE);

	process_root_flag ("-R", argc, argv);

	OPENLOG ("pwconv");

	process_flags (argc, argv);

#ifdef WITH_TCB
	if (getdef_bool("USE_TCB")) {
		fprintf (stderr, _("%s: can't work with tcb enabled\n"), Prog);
		exit (E_FAILURE);
	}
#endif				/* WITH_TCB */

	if (pw_lock () == 0) {
		fprintf (stderr,
		         _("%s: cannot lock %s; try again later.\n"),
		         Prog, pw_dbname ());
		fail_exit (E_PWDBUSY);
	}
	pw_locked = true;
	if (pw_open (O_RDWR) == 0) {
		fprintf (stderr,
		         _("%s: cannot open %s\n"), Prog, pw_dbname ());
		fail_exit (E_MISSING);
	}

	if (spw_lock () == 0) {
		fprintf (stderr,
		         _("%s: cannot lock %s; try again later.\n"),
		         Prog, spw_dbname ());
		fail_exit (E_PWDBUSY);
	}
	spw_locked = true;
	if (spw_open (O_CREAT | O_RDWR) == 0) {
		fprintf (stderr,
		         _("%s: cannot open %s\n"), Prog, spw_dbname ());
		fail_exit (E_FAILURE);
	}

	/*
	 * Remove /etc/shadow entries for users not in /etc/passwd.
	 */
	(void) spw_rewind ();
	while ((sp = spw_next ()) != NULL) {
		if (pw_locate (sp->sp_namp) != NULL) {
			continue;
		}

		if (spw_remove (sp->sp_namp) == 0) {
			/*
			 * This shouldn't happen (the entry exists) but...
			 */
			fprintf (stderr,
			         _("%s: cannot remove entry '%s' from %s\n"),
			         Prog, sp->sp_namp, spw_dbname ());
			fail_exit (E_FAILURE);
		}
	}

	/*
	 * Update shadow entries which don't have "x" as pw_passwd. Add any
	 * missing shadow entries.
	 */
	(void) pw_rewind ();
	while ((pw = pw_next ()) != NULL) {
		sp = spw_locate (pw->pw_name);
		if (NULL != sp) {
			/* do we need to update this entry? */
			if (strcmp (pw->pw_passwd, SHADOW_PASSWD_STRING) == 0) {
				continue;
			}
			/* update existing shadow entry */
			spent = *sp;
		} else {
			/* add new shadow entry */
			memset (&spent, 0, sizeof spent);
			spent.sp_namp   = pw->pw_name;
			spent.sp_min    = getdef_num ("PASS_MIN_DAYS", -1);
			spent.sp_max    = getdef_num ("PASS_MAX_DAYS", -1);
			spent.sp_warn   = getdef_num ("PASS_WARN_AGE", -1);
			spent.sp_inact  = -1;
			spent.sp_expire = -1;
			spent.sp_flag   = SHADOW_SP_FLAG_UNSET;
		}
		spent.sp_pwdp = pw->pw_passwd;
		spent.sp_lstchg = (long) time ((time_t *) 0) / SCALE;
		if (0 == spent.sp_lstchg) {
			/* Better disable aging than requiring a password
			 * change */
			spent.sp_lstchg = -1;
		}
		if (spw_update (&spent) == 0) {
			fprintf (stderr,
			         _("%s: failed to prepare the new %s entry '%s'\n"),
			         Prog, spw_dbname (), spent.sp_namp);
			fail_exit (E_FAILURE);
		}

		/* remove password from /etc/passwd */
		pwent = *pw;
		pwent.pw_passwd = SHADOW_PASSWD_STRING;	/* XXX warning: const */
		if (pw_update (&pwent) == 0) {
			fprintf (stderr,
			         _("%s: failed to prepare the new %s entry '%s'\n"),
			         Prog, pw_dbname (), pwent.pw_name);
			fail_exit (E_FAILURE);
		}
	}

	if (spw_close () == 0) {
		fprintf (stderr,
		         _("%s: failure while writing changes to %s\n"),
		         Prog, spw_dbname ());
		SYSLOG ((LOG_ERR, "failure while writing changes to %s", spw_dbname ()));
		fail_exit (E_FAILURE);
	}
	if (pw_close () == 0) {
		fprintf (stderr,
		         _("%s: failure while writing changes to %s\n"),
		         Prog, pw_dbname ());
		SYSLOG ((LOG_ERR, "failure while writing changes to %s", pw_dbname ()));
		fail_exit (E_FAILURE);
	}

	/* /etc/passwd- (backup file) */
	if (chmod (PASSWD_FILE "-", 0600) != 0) {
		fprintf (stderr,
		         _("%s: failed to change the mode of %s to 0600\n"),
		         Prog, PASSWD_FILE "-");
		SYSLOG ((LOG_ERR, "failed to change the mode of %s to 0600", PASSWD_FILE "-"));
		/* continue */
	}

	if (pw_unlock () == 0) {
		fprintf (stderr, _("%s: failed to unlock %s\n"), Prog, pw_dbname ());
		SYSLOG ((LOG_ERR, "failed to unlock %s", pw_dbname ()));
		/* continue */
	}

	if (spw_unlock () == 0) {
		fprintf (stderr, _("%s: failed to unlock %s\n"), Prog, spw_dbname ());
		SYSLOG ((LOG_ERR, "failed to unlock %s", spw_dbname ()));
		/* continue */
	}

	nscd_flush_cache ("passwd");

	return E_SUCCESS;
}

